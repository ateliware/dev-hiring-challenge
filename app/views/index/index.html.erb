<%= render 'layouts/menu' %>
<div class="container-site">
    <div class="pricing-header px-3 py-3 pt-md-5 pb-md-4 mx-auto text-center">
        <h1 class="display-4">Rpositórios destaques do GitHub</h1>
        <p class="lead">Clique em buscar e veja os repositórios destauqes do GitHub, de acordo com as linguagens de programação mais usadas no mercado, para desenvolvimento WEB.</p>
        <div class="col-md-4 div-center">
            <button type="button" onclick="buscaRepositorios();" class="btn btn-lg btn-block btn-primary btn-outline-primary-red">Buscar</button>
        </div>
    </div>

    <div class="container">
        <div id="contaier_linguagens" class="card-deck mb-3 text-center">
            <div class="card mb-4 box-shadow">
                <div class="card-header">
                    <h4 id="titulo_lista_node" class="my-0 font-weight-normal verde-claro">Node JS</h4>
                </div>
                <div class="card-body">
                    <%= image_tag("logo_nodejs.png", :alt => "Logo node js", :class => "logo-banner-linguagens") %>
                    <ul class="list-unstyled mt-3 mb-4 lista-detalhes-linguagem">
                        <li>Licença: Código Aberto</li>
                        <li>Criação: 27/05/2009</li>
                        <li>Autor: Ryan Dahl</li>
                    </ul>
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-header">
                    <h4 id="titulo_lista_django" class="my-0 font-weight-normal verde-escuro">Django</h4>
                </div>
                <div class="card-body">
                    <%= image_tag("logo_django.png", :alt => "Logo django", :class => "logo-banner-linguagens") %>
                    <ul class="list-unstyled mt-3 mb-4 lista-detalhes-linguagem">
                        <li>Licença: BSD</li>
                        <li>Criação: 21/07/2005</li>
                        <li>Autor: Django Software Foundation</li>
                    </ul>
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-header">
                    <h4 id="titulo_lista_rails" class="my-0 font-weight-normal vermelho">Rails</h4>
                </div>
                <div class="card-body">
                    <%= image_tag("logo_rails.png", :alt => "Logo rails", :class => "logo-banner-linguagens") %>
                    <ul class="list-unstyled mt-3 mb-4 lista-detalhes-linguagem">
                        <li>Licença: MIT</li>
                        <li>Criação: 01/07/2004</li>
                        <li>Autor: Rails Core Team</li>
                    </ul>                    
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-header">
                    <h4 id="titulo_lista_dotnet" class="my-0 font-weight-normal roxo">Dotnet</h4>
                </div>
                <div class="card-body">
                    <%= image_tag("logo_dotnet.png", :alt => "Logo dotnet", :class => "logo-banner-linguagens") %>
                    <ul class="list-unstyled mt-3 mb-4 lista-detalhes-linguagem">
                        <li>Licença: Mista</li>
                        <li>Criação: 13/02/2002</li>
                        <li>Autor: Microsoft</li>
                    </ul>
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-header">
                    <h4 id="titulo_lista_php" class="my-0 font-weight-normal azul">PHP</h4>
                </div>
                <div class="card-body">
                    <%= image_tag("logo_php.png", :alt => "Logo php", :class => "logo-banner-linguagens") %>
                    <ul class="list-unstyled mt-3 mb-4 lista-detalhes-linguagem">
                        <li>Licença: Código Aberto</li>
                        <li>Criação: 1995</li>
                        <li>Autor: Rasmus Lerdorf</li>
                    </ul>
                </div>
            </div>
        </div>
        <div id="contaier_listas" class="card-deck mb-3 text-center oculto">
            <div class="card mb-4 box-shadow">
                <div class="card-body">
                    <button type="button" onclick="detalhes('lista_node');" class="btn btn-lg btn-block btn-primary btn-outline-primary">Detalhes</button>
                    <ul id="lista_node" class="list-unstyled mt-3 mb-4 lista-linguagens"></ul>
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-body">
                    <button type="button" onclick="detalhes('lista_django');" class="btn btn-lg btn-block btn-primary btn-outline-primary">Detalhes</button>
                    <ul id="lista_django" class="list-unstyled mt-3 mb-4 lista-linguagens"></ul>
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-body">
                    <button type="button" onclick="detalhes('lista_rails');" class="btn btn-lg btn-block btn-primary btn-outline-primary">Detalhes</button>
                    <ul id="lista_rails" class="list-unstyled mt-3 mb-4 lista-linguagens"></ul>
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-body">
                    <button type="button" onclick="detalhes('lista_dotnet');" class="btn btn-lg btn-block btn-primary btn-outline-primary">Detalhes</button>
                    <ul id="lista_dotnet" class="list-unstyled mt-3 mb-4 lista-linguagens"></ul>
                </div>
            </div>
            <div class="card mb-4 box-shadow">
                <div class="card-body">
                    <button type="button" onclick="detalhes('lista_php');" class="btn btn-lg btn-block btn-primary btn-outline-primary">Detalhes</button>
                    <ul id="lista_php" class="list-unstyled mt-3 mb-4 lista-linguagens"></ul>
                </div>
            </div>
        </div>
        <div class="card-deck mb-3 text-center">
            <ul id="lista_repositorios"></ul>
        </div>
        <%= render 'layouts/footer' %>
    </div>
</div>

<div class="modal fade" id="modalDetalhes" tabindex="-1" role="dialog" aria-labelledby="modalDetalhesLabel" aria-hidden="true" data-target=".bd-example-modal-lg">
    <div class="modal-dialog modal-lg" role="document">
        <div class="modal-content">
            <div class="modal-header">
                <h5 class="modal-title" id="modalDetalhesLabel"></h5>
                <button type="button" class="close" data-dismiss="modal" aria-label="Close">
                    <span class="vermelho" aria-hidden="true">&times;</span>
                </button>
            </div>
            <div class="modal-body corpo-modal">
                <ul id="lista_linguagens_detalhada" class="list-container">

                </ul>
            </div>
            <div class="modal-footer">
                <button type="button" class="btn btn-secondary btn-outline-primary-red" data-dismiss="modal">OK</button>
            </div>
        </div>
    </div>
</div>
