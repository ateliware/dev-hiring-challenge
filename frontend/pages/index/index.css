* {
    font-family: -apple-system, BlinkMacSystemFont, “Segoe UI”, Roboto, Oxygen-Sans, Ubuntu, Cantarell, Helvetica, Arial, sans-serif !important;
    text-align: center;
}

a {
    text-decoration: none !important;
}

a:hover {
    text-decoration: underline !important;
}

select {
    text-align: start !important;
    height: calc(3.5rem + 10px) !important;
}

option,
.toast-body {
    text-align: start !important;
}

.toast {
    border: 0px !important;
}

.toast-title-icon::before {
    display: inline-block;
    min-height: 12px;
    min-width: 12px;
    margin-right: 4px;
    margin-bottom: -1px;
    border-radius: 8px;
    content: "";
}

.toast-title-icon.success::before {
    background: green;
}

.toast-title-icon.error::before {
    background: red;
}

th.col-actions {
    width: 80px !important;
}

td.col-actions {
    width: 80px !important;
    display: flex;
    justify-content: center;
    align-items: center;
}

.col-actions button {
    border: none;
    padding: 3px 6px;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 4px;
}

.modal-body .row div {
    display: flex;
    justify-content: flex-start;

}

.modal-body .row div b {
    margin-right: 4px;
}

/***** Loader *****/
body.loader {
    overflow: hidden !important;
}

div.loader {
    position: fixed;
    top: 0;
    left: 0;
    background: rgb(0 0 0 / 50%);
    z-index: 99996;
    width: 100vw;
    height: 100vh;
}

div.spinner-wrapper {
    border-radius: 8px;
    z-index: 99997;
    display: block;
    position: fixed;
    background-color: #FFFFFF;
    height: 130px;
    width: 130px;
    top: calc(50% - 65px);
    left: calc(50% - 65px);
}

p.spinner-label {
    -webkit-user-select: none;
    user-select: none;
    z-index: 99998;
    margin-top: 97px;
    text-align: center;
}

div.spinner {
    display: block;
    position: fixed;
    z-index: 99998;
    top: calc(50% - 42px);
    left: calc(50% - 30px);
    height: 60px;
    width: 60px;
    border: solid 4px transparent;
    border-top-color: #000000;
    border-left-color: #000000;
    border-radius: 100px;
    -webkit-animation: loader-spinner 1000ms linear infinite;
    -moz-animation: loader-spinner 1000ms linear infinite;
    -ms-animation: loader-spinner 1000ms linear infinite;
    -o-animation: loader-spinner 1000ms linear infinite;
    animation: loader-spinner 1000ms linear infinite;
}

@-webkit-keyframes loader-spinner {
    0% {
        -webkit-transform: rotate(0deg);
        transform: rotate(0deg);
    }
    100% {
        -webkit-transform: rotate(360deg);
        transform: rotate(360deg);
    }
}

@-moz-keyframes loader-spinner {
    0% {
        -moz-transform: rotate(0deg);
        transform: rotate(0deg);
    }
    100% {
        -moz-transform: rotate(360deg);
        transform: rotate(360deg);
    }
}

@-o-keyframes loader-spinner {
    0% {
        -o-transform: rotate(0deg);
        transform: rotate(0deg);
    }
    100% {
        -o-transform: rotate(360deg);
        transform: rotate(360deg);
    }
}

@-ms-keyframes loader-spinner {
    0% {
        -ms-transform: rotate(0deg);
        transform: rotate(0deg);
    }
    100% {
        -ms-transform: rotate(360deg);
        transform: rotate(360deg);
    }
}

@keyframes loader-spinner {
    0% {
        transform: rotate(0deg);
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
        transform: rotate(360deg);
    }
}

/***** Mobile fixes *****/

@media (max-width: 576px) {
    .col-sequencial-number,
    .col-actions {
        display: none;
    }
    .fade {
        opacity: 1px !important;
    }
    .table>:not(caption)>*>* {
        padding: 0.5rem 0.3rem !important;
        font-size: 14px !important;
    }
    select {
        text-align: center !important;
        height: calc(3.5rem + 25px) !important;
    }
    option {
        text-align: center !important;
    }
    .form-floating>.form-select {
        padding-top: 2.925rem !important
    }
}