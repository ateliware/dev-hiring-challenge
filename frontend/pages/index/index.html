<!DOCTYPE html>
<html lang="pt-BR">

<head>

    <!-- General -->
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description" content="Veja os melhores repositórios das melhores linguagens de programação!">
    <meta name="author" content="Eduardo Mior">
    <meta name="robots" content="noindex, nofollow">
    <title>Top GitHub Repositórios 😎</title>

    <!-- Icons -->
    <link rel="icon" type="image/png" href="icon.png" sizes="192x192">
    <link rel="apple-touch-icon" type="image/png" href="apple-touch-icon.png" sizes="180x180">

    <!-- Android -->
    <meta name="theme-color" content="#FFFFFF">
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="application-name" content="Top GitHub Repositórios 😎">

    <!-- iOS -->
    <meta name="apple-mobile-web-app-title" content="Top GitHub Repositórios 😎">
    <meta name="apple-mobile-web-app-capable" content="yes">

    <!-- OG -->
    <meta property="og:locale" content="pt_BR">
    <meta property="og:type" content="website">
    <meta property="og:title" content="Top GitHub Repositórios 😎">
    <meta property="og:site_name" content="Top GitHub Repositórios 😎">
    <meta property="og:description" content="Veja os melhores repositórios das melhores linguagens de programação!">
    <meta property="og:url" content="https://url-aqui.com">
    <meta property="og:image" content="https://url-aqui.com/images/logo2.png">
    <meta property="og:image:type" content="image/png">
    <meta property="og:image:alt" content="Top GitHub Repositórios 😎">

    <!-- CSS -->
    <link rel="stylesheet" href="index.css" />
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css">

    <!-- JS -->
    <script src="https://browser.sentry-cdn.com/7.1.1/bundle.min.js"></script>
    <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/js/bootstrap.bundle.min.js"></script>
    <script src="index.js"></script>

</head>

<body>
    <div class="container">

        <div class="row d-flex justify-content-center">

            <div class="col-12 mt-3">
                <h1 class="mb-2">Top GitHub Repositórios 😎</h1>
            </div>

            <div class="col-12">
                <span>Veja em tempo real os melhores repositórios das melhores linguagens de programação da atualidade!</span>
            </div>

            <div class="col-12 d-flex justify-content-center mb-1 mt-3 mt-sm-5">
                <div class="form-floating col-8">
                    <select class="form-select form-select-lg" id="linguagem-programacao">
                        <option selected disabled></option>
                        <option value="java">☕ Java</option>
                        <option value="ocaml">🐫 OCaml</option>
                        <option value="lua">🌙 Lua</option>
                        <option value="python">🐍 Python</option>
                        <option value="ruby">💎 Ruby</option>
                    </select>
                    <label for="linguagem-programacao">Selecione a sua linguagem de programação favorita</label>
                </div>
            </div>

            <div class="col-12 mt-1 mt-sm-4" id="dados-linguagem" style="display: none;">
                <div class="row d-flex justify-content-between">
                    <div class="col-12 col-md-4 mt-3">
                        <button id="buscar-repositorios-github" type="button" class="col-12 btn btn-success"
                            title="Clique para buscar no GitHub a lista atualizada dos TOP repositórios da sua linguagem de programação favorita">🐙 Buscar dados no GitHub</button>
                    </div>
                    <div class="col-12 col-md-4 mt-3">
                        <button id="salvar-repositorios-em-cache" type="button" class="col-12 btn btn-primary" disabled
                            title="Clique para salvar no banco de dados a lista de repositórios buscados no GitHub">💾 Salvar dados no DB</button>
                    </div>
                    <div class="col-12 col-md-4 mt-3">
                        <button id="excluir-repositorios-do-cache" type="button" class="col-12 btn btn-danger" disabled
                            title="Clique para apagar do banco de dados toda a lista de repositórios buscados no GitHub">🗑 Limpar dados do DB</button>
                    </div>
                </div>
                <table class="table mt-1">
                    <thead>
                        <tr>
                            <th scope="col" class="col-sequencial-number">#</th>
                            <th scope="col" class="col-repository">Repositório</th>
                            <th scope="col" class="col-owner">Owner</th>
                            <th scope="col" class="col-creation-date">Data de Criação</th>
                            <th scope="col" class="col-stars">Stars</th>
                            <th scope="col" class="col-actions">Detalhes</th>
                        </tr>
                    </thead>
                    <tbody class="table-group-divider">
                    </tbody>
                </table>
            </div>

        </div>

    </div>
</body>

</html>
