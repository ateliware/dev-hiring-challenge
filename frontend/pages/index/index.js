function handleError() {
    hideLoader();
}

$(document).ready(function () {

    $("#linguagem-programacao").on("change", function () {
        // TODO
    });

    $("#buscar-repositorios").on("click", function () {
        let linguagemSelecionada = $("#linguagem-programacao").val();


        $.get(`https://api.github.com/search/repositories?sort=stars&per_page=5&q=language:${linguagemSelecionada}`)
            .fail(handleError)
            .done(function (result) {
                $("table tbody tr").remove();

                result.items.forEach((repositorio, i) => {
                    let tableRow = $("<tr/>");
                    tableRow.append(`<td>${i + 1}</td>`);
                    tableRow.append(`<td>${repositorio.name}</td>`);
                    tableRow.append(`
                                <td>
                                    <span title="O owner deste reposit√≥rio √© ${repositorio.owner.type == "Organization" ? "uma organiza√ß√£o" : "um usu√°rio"}">${repositorio.owner.type == "Organization" ? "üè¢" : "üë§"}</span> 
                                    <a href="${repositorio.owner.html_url}" target="_blank" title="Clique para visitar o perfil">${repositorio.owner.login}</a>
                                </td>`);
                    tableRow.append(`<td>${new Date(repositorio.created_at).toLocaleDateString()}</td>`);
                    tableRow.append(`<td>${repositorio.stargazers_count}</td>`);
                    tableRow.append(`<td></td>`);
                    $("table tbody").append(tableRow);
                });

            });
    });

    $("#salvar-repositorios-em-cache").on("click", function () {
        // TODO
    });

    $("#excluir-repositorios-do-cache").on("click", function () {
        // TODO
    });

});
