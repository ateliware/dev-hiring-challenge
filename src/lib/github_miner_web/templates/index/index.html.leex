<div class="container mx-auto px-4 md:container md:mx-auto">

      <div class="text-center mb-5 mt-5">
        <p class="text-3xl text-cyan-700 font-bold">Escolha uma linguagem de programação</p>
      </div>

      <div class="text-center mb-5 mt-5">
          <button phx-click="search-repo" phx-value-lang="elixir"> Elixir </button>
          <button phx-click="search-repo" phx-value-lang="python"> Python </button>
          <button phx-click="search-repo" phx-value-lang="php"> PHP </button>
          <button phx-click="search-repo" phx-value-lang="java"> Java </button>
          <button phx-click="search-repo" phx-value-lang="ruby"> Ruby </button>
      </div>

      <%= live_component(@socket, GithubMinerWeb.ShowRepositoriesComponent, id: "repositories", list_repos: @list_repos, search: @search) %>

      <%= if @show_more do %>
        <%= live_component(@socket, GithubMinerWeb.ShowMoreComponent, id: "repositories", list_repos_full: @list_repos_full, show_more: @show_more, list_repos: @list_repos) %>
      <% end %>

</div>
