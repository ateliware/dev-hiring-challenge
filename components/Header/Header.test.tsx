import { render, screen } from "@testing-library/react";
import { Header } from "./index";

describe("Header", () => {
  it("renderiza o titulo do aplicativo", () => {
    render(<Header />);
    const pageTitle = screen.getByRole("heading", {
      name: "Prova Ateliware",
    });
    expect(pageTitle).toBeInTheDocument();
  });
  it("renderiza email do autor", () => {
    const { container } = render(<Header />);
    const email = screen.getByText(new RegExp("HOLA@FACUNDOLEITES.COM", "gi"));
    expect(email).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
