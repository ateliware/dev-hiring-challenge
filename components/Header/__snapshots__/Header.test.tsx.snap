// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renderiza email do autor 1`] = `
<div>
  <header>
    <section
      class="container"
    >
      <div
        class="main"
      >
        <h1
          class="title"
        >
          Prova Ateliware
        </h1>
        <p
          class="lead"
        >
          Desafio t√©cnico para desenvolvedores
        </p>
        <section
          class="userInfoSectionContainer"
        >
          <div
            class="main"
          >
            <hr
              class="divider"
            />
            <a
              class="mailToButton"
              href="mailto:hola@facundoleites.com"
            >
              hola@facundoleites.com
            </a>
          </div>
        </section>
      </div>
    </section>
  </header>
</div>
`;
